File: JANUARY/14-01-2026/3454.swift
class Solution {
    func separateSquares(_ squares: [[Int]]) -> Double {
        let n = squares.count

        var xs = [Int64]()
        xs.reserveCapacity(2 * n)

        struct Event {
            let y: Int64
            let x1: Int64
            let x2: Int64
            let t: Int
        }

        var events = [Event]()
        events.reserveCapacity(2 * n)

        for s in squares {
            let x = Int64(s[0])
            let y = Int64(s[1])
            let l = Int64(s[2])
            xs.append(x)
            xs.append(x + l)
            events.append(Event(y: y, x1: x, x2: x + l, t: 1))
            events.append(Event(y: y + l, x1: x, x2: x + l, t: -1))
        }

        xs.sort()
        var uniq = [Int64]()
        for v in xs {
            if uniq.isEmpty || uniq.last! != v {
                uniq.append(v)
            }
        }

        let m = uniq.count
        var index = [Int64: Int]()
        for i in 0..<m {
            index[uniq[i]] = i
        }

        var cnt = Array(repeating: Int64(0), count: 4 * m)
        var seglen = Array(repeating: Int64(0), count: 4 * m)

        func update(_ i: Int, _ l: Int, _ r: Int, _ ql: Int, _ qr: Int, _ v: Int) {
            if qr <= l || r <= ql { return }
            if ql <= l && r <= qr {
                cnt[i] += Int64(v)
            } else {
                let mid = (l + r) >> 1
                update(i << 1, l, mid, ql, qr, v)
                update(i << 1 | 1, mid, r, ql, qr, v)
            }
            if cnt[i] > 0 {
                seglen[i] = uniq[r] - uniq[l]
            } else if l + 1 == r {
                seglen[i] = 0
            } else {
                seglen[i] = seglen[i << 1] + seglen[i << 1 | 1]
            }
        }

        events.sort { $0.y < $1.y }

        var prevY = events[0].y
        var total = 0.0
        var i = 0

        while i < events.count {
            let y = events[i].y
            let dy = y - prevY
            if dy > 0 && seglen[1] > 0 {
                total += Double(seglen[1]) * Double(dy)
            }
            while i < events.count && events[i].y == y {
                let e = events[i]
                update(1, 0, m, index[e.x1]!, index[e.x2]!, e.t)
                i += 1
            }
            prevY = y
        }

        let half = total / 2.0
        cnt = Array(repeating: 0, count: 4 * m)
        seglen = Array(repeating: 0, count: 4 * m)

        prevY = events[0].y
        var cur = 0.0
        i = 0

        while i < events.count {
            let y = events[i].y
            let dy = y - prevY
            if dy > 0 && seglen[1] > 0 {
                let area = Double(seglen[1]) * Double(dy)
                if cur + area >= half {
                    return Double(prevY) + (half - cur) / Double(seglen[1])
                }
                cur += area
            }
            while i < events.count && events[i].y == y {
                let e = events[i]
                update(1, 0, m, index[e.x1]!, index[e.x2]!, e.t)
                i += 1
            }
            prevY = y
        }

        return Double(prevY)
    }
}

File: JANUARY/14-01-2026/3454.cs
public class Solution {
    public double SeparateSquares(int[][] squares) {
        int n = squares.Length;
        long[] xs = new long[2 * n];
        long[][] events = new long[2 * n][];
        for (int i = 0; i < n; i++) {
            long x = squares[i][0], y = squares[i][1], l = squares[i][2];
            xs[2 * i] = x;
            xs[2 * i + 1] = x + l;
            events[2 * i] = new long[] { y, x, x + l, 1 };
            events[2 * i + 1] = new long[] { y + l, x, x + l, -1 };
        }

        Array.Sort(xs);
        int m = 0;
        for (int i = 0; i < xs.Length; i++) {
            if (i == 0 || xs[i] != xs[i - 1]) xs[m++] = xs[i];
        }

        Array.Sort(events, (a, b) => a[0].CompareTo(b[0]));

        long[] cnt = new long[4 * m];
        long[] len = new long[4 * m];

        void Update(int i, int l, int r, int ql, int qr, long v) {
            if (qr <= l || r <= ql) return;
            if (ql <= l && r <= qr) {
                cnt[i] += v;
            } else {
                int mid = (l + r) >> 1;
                Update(i << 1, l, mid, ql, qr, v);
                Update(i << 1 | 1, mid, r, ql, qr, v);
            }
            if (cnt[i] > 0) len[i] = xs[r] - xs[l];
            else if (l + 1 == r) len[i] = 0;
            else len[i] = len[i << 1] + len[i << 1 | 1];
        }

        long prevY = events[0][0];
        double total = 0;
        int idxEv = 0;

        while (idxEv < events.Length) {
            long y = events[idxEv][0];
            long dy = y - prevY;
            if (dy > 0 && len[1] > 0) total += len[1] * dy;
            while (idxEv < events.Length && events[idxEv][0] == y) {
                int l = Array.BinarySearch(xs, 0, m, events[idxEv][1]);
                int r = Array.BinarySearch(xs, 0, m, events[idxEv][2]);
                Update(1, 0, m - 1, l, r, events[idxEv][3]);
                idxEv++;
            }
            prevY = y;
        }

        double half = total / 2.0;
        Array.Clear(cnt, 0, cnt.Length);
        Array.Clear(len, 0, len.Length);

        prevY = events[0][0];
        double cur = 0;
        idxEv = 0;

        while (idxEv < events.Length) {
            long y = events[idxEv][0];
            long dy = y - prevY;
            if (dy > 0 && len[1] > 0) {
                double area = len[1] * dy;
                if (cur + area >= half)
                    return prevY + (half - cur) / len[1];
                cur += area;
            }
            while (idxEv < events.Length && events[idxEv][0] == y) {
                int l = Array.BinarySearch(xs, 0, m, events[idxEv][1]);
                int r = Array.BinarySearch(xs, 0, m, events[idxEv][2]);
                Update(1, 0, m - 1, l, r, events[idxEv][3]);
                idxEv++;
            }
            prevY = y;
        }

        return prevY;
    }
}